### PR TITLE
Feat: add servercow provider

### PR DESCRIPTION
**Pull request by baswag**

### CODE REVIEW

File: internal/settings/providers/servercow/provider.go [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+
+func (p *provider) Update(ctx context.Context, client *http.Client, ip net.IP) (newIP net.IP, err error) {
+	recordType := constants.A
+	if ip.To4() == nil {
+		recordType = constants.AAAA
+	}
+	u := url.URL{
+		Scheme: "https",
+		Host:   "api.servercow.de",
+		Path:   fmt.Sprintf("/dns/v1/domains/%s", p.domain),
```
Comment (qdm12) [inline internal/settings/providers/servercow/provider.go:112]:
```suggestion
		Path:   "/dns/v1/domains/" + p.domain,
```

File: internal/settings/providers/servercow/provider.go [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+	}
+	values := url.Values{}
+	u.RawQuery = values.Encode()
+	if !p.useProviderIP {
+		if ip.To4() == nil {
+			values.Set("ip6", ip.String())
+		} else {
+			values.Set("ip", ip.String())
+		}
+	}
```
Comment (qdm12) [inline internal/settings/providers/servercow/provider.go:122]: You are setting the ip url value at the wrong place 😉 

```suggestion
	values := url.Values{}
	if !p.useProviderIP {
		if ip.To4() == nil {
			values.Set("ip6", ip.String())
		} else {
			values.Set("ip", ip.String())
		}
	}
	u.RawQuery = values.Encode()
```

File: internal/settings/providers/servercow/provider.go [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+	requestData := struct {
+		Type    string `json:"type"`    // constants.A or constants.AAAA depending on ip address given
+		Name    string `json:"name"`    // DNS record name i.e. example.com
+		Content string `json:"content"` // ip address
+		TTL     uint   `json:"ttl"`
+	}{
+		Type:    recordType,
+		Name:    p.host,
```
Comment (qdm12) [inline internal/settings/providers/servercow/provider.go:131]: this should be domain or p.BuildDomainName() (to have host.domain.com for example)

```suggestion
		Name:    p.domain,
```

File: internal/settings/providers/servercow/provider.go [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+	var parsedJSON struct {
+		Message string `json:"message"`
+		Error string `json:"error"`
+	}
+
+	if err := decoder.Decode(&parsedJSON); err != nil {
+		return nil, fmt.Errorf("%w: %s", errors.ErrUnmarshalResponse, err)
+	}
+
+	if(parsedJSON.Message != "ok") {
```
Comment (qdm12) [inline internal/settings/providers/servercow/provider.go:175]:
```suggestion
	if parsedJSON.Message != "ok" {
```

File: internal/settings/providers/servercow/provider.go [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+		return nil, fmt.Errorf("%w: %d: %s",
+			errors.ErrBadHTTPStatus, response.StatusCode, utils.BodyToSingleLine(response.Body))
+	}
+
+	decoder := json.NewDecoder(response.Body)
+
+	var parsedJSON struct {
+		Message string `json:"message"`
+		Error string `json:"error"`
+	}
```
Comment (qdm12) [inline internal/settings/providers/servercow/provider.go:169]: Nice 👍

File: internal/settings/settings.go [hunk 6]
```diff
@@ -110,6 +111,8 @@ func New(provider models.Provider, data json.RawMessage, domain, host string,
 		return ovh.New(data, domain, host, ipVersion, logger)
 	case constants.SelfhostDe:
 		return selfhostde.New(data, domain, host, ipVersion, logger)
+	case constants.Servercow:
+		return servercow.New(data,domain,host,ipVersion,logger)
```
Comment (qdm12) [inline internal/settings/settings.go:115]: this needs formatting as well

**Review by qdm12 — COMMENTED** [review]
Thanks for the contribution!!

A few small changes that have to be fixed before I can merge that as follow.

File: internal/settings/providers/servercow/provider.go (hunk 2 above)
Comment (baswag) [reply internal/settings/providers/servercow/provider.go:122]: Accidently left those in, removed unneeded code

File: internal/settings/providers/servercow/provider.go (hunk 3 above)
Comment (baswag) [reply internal/settings/providers/servercow/provider.go:131]: Updated comment to make clear that only the subdomain part is accepted by the Servercow API

File: internal/settings/providers/servercow/provider.go [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+	host          string
+	domain        string
+	ipVersion     ipversion.IPVersion
+	password      string
+	useProviderIP bool
+	ttl           uint
+	logger        log.Logger
+}
+
+func New(data json.RawMessage, domain, host string, ipVersion ipversion.IPVersion, logger log.Logger) (p *provider, err error) {
```
Comment (qdm12) [inline internal/settings/providers/servercow/provider.go:33]: The CI is failing due to linting as the 'line is too long', you can just commit that suggestion and it *should* pass

```suggestion
func New(data json.RawMessage, domain, host string, ipVersion ipversion.IPVersion,
	logger log.Logger) (p *provider, err error) {
```

**Review by qdm12 — APPROVED** [review]
Nice thanks for the modifications and fixes! Just one tiny more fix that makes the CI fail and we can merge this!

**Comment by qdm12** [comment]
Nice thanks, it's merged now. I pushed that tiny a4af81dddb58854b5379654d37d1d3a3a2cdfe95 commit so that the lint stage passes as there 2 more linting errors. It should now be incorporated in the `:latest` image. I'll do a release image tag sometime soon.

**Pull request merged**